README.md
# automate_alert

A new Flutter package which helps developers create automated closed pop up alert
 
# Screen

<img src="https://github.com/Eva1234ev/AutomateAlert/blob/master/demo1.gif?raw=true" width="270"> <img src="https://github.com/Eva1234ev/AutomateAlert/blob/master/demo2.gif?raw=true" width="270">


## Features

- Adding buttons dynamically (as much as you want)
- Predefined beautiful alert styles (success, error, warning, info)
- Set animation duration
- Reusable alert styles

## Usage

[Example](https://github.com/Eva1234ev/AutomateAlert/edit/blob/master/example/example.dart)

TO use this plugin:
 * add dependency to your [pabspec.yaml] (https://github.com/Eva1234ev/AutomateAlert/edit/blob/master/pubspec.yaml)

```yaml
dependencies:
  flutter:
    sdk: flutter
    automate_alert:
```


### How to use?

```dart
//Simple error buton to show auto alert 
FloatingActionButton.extended(
                  onPressed: () {
                    AlertNoServerAutoDialog(
                        context: context,
                        message: "Can not connect server!",
                        subMessage:
                            'Please try again and check your internet connection',
                        alertType: AlertType.Error,
                        showDuration: 4)
                      ..show();
                  },
                  label: Text('Error Alert',
                      style: TextStyle(
                        color: Colors.white,
                      )),
                  icon: Icon(Icons.error, color: Colors.white),
                  backgroundColor: Colors.redAccent,
                ),
```
# License 

Copyright (c) 2021 Eva

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
## [0.0.1] - Release

* A customizable automated closed pop up alert.


lib/automate_alert.dart
library automate_alert;

import 'dart:async';
import 'package:flutter/material.dart';
import 'alert_type_attribute.dart';
import 'alert_slow_show.dart';
import 'alert_Icon_show.dart';

enum AlertType { Error, Success, Warning, Info }
enum AlertAnimateType { SizeAnimate, SlowAppear }

class AlertNoServerAutoDialog {
  final BuildContext context;

  final String message;

  final String subMessage;

  final AlertType alertType;

  final AlertAnimateType alertAnimateType;

  final int showDuration;

  AlertNoServerAutoDialog(
      {@required this.context,
      this.message,
      this.subMessage = '',
      this.alertType = AlertType.Error,
      this.alertAnimateType = AlertAnimateType.SlowAppear,
      this.showDuration = 2});

  Timer _timer;
  AlertTypeAttribute alertTypeAttribute;
  Future show() => showDialog(
          context: context,
          builder: (BuildContext builderContext) {
            _getIconResult(alertType);
            _timer = Timer(Duration(seconds: showDuration), () {
              Navigator.of(context).pop();
            });
            return AlertDialog(
                content: Column(
                    mainAxisSize: MainAxisSize.min,
                    crossAxisAlignment: CrossAxisAlignment.stretch,
                    children: <Widget>[
                  _getAnimate(alertAnimateType),
                  SlowShow(
                    child: Text(message,
                        maxLines: 4,
                        textAlign: TextAlign.center,
                        style: new TextStyle(
                          fontSize: 16.0,
                          fontFamily: 'Roboto',
                          color: alertTypeAttribute.alertColor,
                          fontWeight: FontWeight.bold,
                        )),
                  ),
                  Container(height: 10),
                  subMessage.isNotEmpty
                      ? SlowShow(
                          child: Text(
                          subMessage,
                          maxLines: 4,
                          textAlign: TextAlign.center,
                          style: new TextStyle(
                            fontSize: 14.0,
                            color: new Color(0xFF212121),
                          ),
                        ))
                      : Container(),
                ]));
          }).then((val) {
        if (_timer.isActive) {
          _timer.cancel();
        }
      });

  void _getIconResult(AlertType type) {
    switch (type) {
      case AlertType.Error:
        alertTypeAttribute = AlertTypeAttribute.fromJson(
            {"alertIcon": Icons.error, "alertColor": Colors.redAccent});
        break;
      case AlertType.Success:
        alertTypeAttribute = AlertTypeAttribute.fromJson(
            {"alertIcon": Icons.check_circle, "alertColor": Colors.green});

        break;
      case AlertType.Info:
        alertTypeAttribute = AlertTypeAttribute.fromJson({
          "alertIcon": Icons.help_outlined,
          "alertColor": Color(0xFF5C6BC0)
        });

        break;
      case AlertType.Warning:
        alertTypeAttribute = AlertTypeAttribute.fromJson(
            {"alertIcon": Icons.info, "alertColor": Color(0xFFF9A825)});

        break;
      default:
        alertTypeAttribute = AlertTypeAttribute.fromJson({
          "alertIcon": Icons.account_circle,
          "alertColor": Color(0xFFEA80FC)
        });
    }
  }

  Widget _getAnimate(AlertAnimateType type) {
    switch (type) {
      case AlertAnimateType.SizeAnimate:
        return AlertIconWidget(alertTypeAttribute);
        break;
      case AlertAnimateType.SlowAppear:
        return SlowShow(
          child: Icon(alertTypeAttribute.alertIcon,
              color: alertTypeAttribute.alertColor, size: 40),
        );
        break;
    }
  }
}


lib/alert_Icon_show.dart
import 'package:flutter/material.dart';
import 'alert_type_attribute.dart';

// ignore: must_be_immutable
class AlertIconWidget extends StatefulWidget {
  AlertTypeAttribute alertAttribute;

  AlertIconWidget(this.alertAttribute);
  @override
  _AlertIconWidgetState createState() => _AlertIconWidgetState();
}

class _AlertIconWidgetState extends State<AlertIconWidget>
    with TickerProviderStateMixin {
  AnimationController controller;

  void initState() {
    super.initState();
    controller = AnimationController(
      vsync: this,
      duration: Duration(seconds: 1),
      lowerBound: 0,
      upperBound: 50,
    );

    controller.addListener(() {
      setState(() {});
    });
    controller.repeat(reverse: true);
  }

  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Icon(widget.alertAttribute.alertIcon,
        color: widget.alertAttribute.alertColor, size: controller.value);
  }
}


lib/alert_slow_show.dart
import 'dart:async';
import 'package:flutter/material.dart';

class SlowShow extends StatefulWidget {
  final Widget child;
  final int delay;

  SlowShow({@required this.child, this.delay});

  @override
  _SlowShowState createState() => _SlowShowState();
}

class _SlowShowState extends State<SlowShow> with TickerProviderStateMixin {
  AnimationController _animController;
  Animation<Offset> _animOffset;

  @override
  void initState() {
    super.initState();

    _animController =
        AnimationController(vsync: this, duration: Duration(milliseconds: 500));
    final curve =
        CurvedAnimation(curve: Curves.decelerate, parent: _animController);
    _animOffset =
        Tween<Offset>(begin: const Offset(0.0, 0.35), end: Offset.zero)
            .animate(curve);

    if (widget.delay == null) {
      _animController.forward();
    } else {
      Timer(Duration(milliseconds: widget.delay), () {
        _animController.forward();
      });
    }
  }

  @override
  void dispose() {
    super.dispose();
    _animController.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return FadeTransition(
      child: SlideTransition(
        position: _animOffset,
        child: widget.child,
      ),
      opacity: _animController,
    );
  }
}


lib/alert_text_show.dart
import 'dart:async';
import 'package:flutter/material.dart';

class TextShow extends StatefulWidget {
  final Widget child;
  final int delay;

  TextShow({@required this.child, this.delay});

  @override
  _TextShowState createState() => _TextShowState();
}

class _TextShowState extends State<TextShow> with TickerProviderStateMixin {
  AnimationController _animController;
  Animation<Offset> _animOffset;

  @override
  void initState() {
    super.initState();

    _animController =
        AnimationController(vsync: this, duration: Duration(milliseconds: 500));
    final curve =
        CurvedAnimation(curve: Curves.decelerate, parent: _animController);
    _animOffset =
        Tween<Offset>(begin: const Offset(0.0, 0.35), end: Offset.zero)
            .animate(curve);

    if (widget.delay == null) {
      _animController.forward();
    } else {
      Timer(Duration(milliseconds: widget.delay), () {
        _animController.forward();
      });
    }
  }

  @override
  void dispose() {
    super.dispose();
    _animController.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return FadeTransition(
      child: SlideTransition(
        position: _animOffset,
        child: widget.child,
      ),
      opacity: _animController,
    );
  }
}


example/example_app.dart
import 'package:flutter/material.dart';
import 'package:automate_alert/automate_alert.dart';

class ShowAlert extends StatefulWidget {
  @override
  _ShowAlertState createState() => _ShowAlertState();
}

class _ShowAlertState extends State<ShowAlert> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text("Auto Close popup"),
          backgroundColor: Color(0xFF90CAF9),
        ),
        body: Container(
            alignment: Alignment.centerLeft,
            padding: EdgeInsets.only(top: 40, left: 40, right: 40, bottom: 10),
            child: ListView(
              padding: const EdgeInsets.all(8),
              children: <Widget>[
                FloatingActionButton.extended(
                  onPressed: () {
                    AlertNoServerAutoDialog(
                        context: context,
                        message: "Can not connect server!",
                        subMessage:
                            'Please try again and check your internet connection',
                        alertType: AlertType.Error,
                        showDuration: 4)
                      ..show();
                  },
                  label: Text('Error Alert',
                      style: TextStyle(
                        color: Colors.white,
                      )),
                  icon: const Icon(Icons.error, color: Colors.white),
                  backgroundColor: Colors.redAccent,
                ),
                Container(
                  height: 15,
                ),
                FloatingActionButton.extended(
                  onPressed: () {
                    AlertNoServerAutoDialog(
                        context: context,
                        message: "Thanks for letting us know!",
                        subMessage:
                            'Your feedback improves the quality of our product',
                        alertType: AlertType.Success,
                        showDuration: 4)
                      ..show();
                  },
                  hoverColor: Colors.grey,
                  label: const Text(
                    'Success Alert',
                    style: TextStyle(
                      color: Colors.white,
                    ),
                  ),
                  icon: const Icon(Icons.check_circle, color: Colors.white),
                  backgroundColor: Colors.green,
                ),
                Container(
                  height: 15,
                ),
                FloatingActionButton.extended(
                  onPressed: () {
                    ScaffoldMessenger.of(context)
                        .showSnackBar(SnackBar(content: Text('Tap')));
                    AlertNoServerAutoDialog(
                      context: context,
                      message: "Low Battery!",
                      subMessage:
                          'Battery temperature too low, your device maybe shut down',
                      alertType: AlertType.Warning,
                    )..show();
                  },
                  label: const Text(
                    'Warning Alert',
                    style: TextStyle(
                      color: Colors.white,
                    ),
                  ),
                  icon: const Icon(Icons.info, color: Colors.white),
                  backgroundColor: Colors.amber[500],
                ),
                Container(
                  height: 15,
                ),
                FloatingActionButton.extended(
                  onPressed: () {
                    AlertNoServerAutoDialog(
                        context: context,
                        message: "App permissions!",
                        subMessage:
                            'Theres nothing else you need to do, except to learn the app running and keep Bluetooth on',
                        alertType: AlertType.Info,
                        showDuration: 4)
                      ..show();
                  },
                  label: const Text(
                    'Info Alert',
                    style: TextStyle(
                      color: Colors.white,
                    ),
                  ),
                  icon: const Icon(Icons.help_outlined, color: Colors.white),
                  backgroundColor: Colors.blue,
                ),
              ],
            )));
  }
}
